src/contrib/mlir/core/pass/ng_dialect_builder.cpp

```diff
@@ -0,0 +1,538 @@
+
+        void createReturn();
+
+        /// Converts nGraph shape-like types \p ng_shape to MLIR shape \p mlir_shape.
+        template <typename T>
+        void getMlirShape(T ngShape, llvm::SmallVectorImpl<int64_t>& mlirShape);
+
+        /// Converts an ngraph shape to an I64 array attribute
+        template <typename T>
+        mlir::ArrayAttr getShapeAsAttr(T ngShape);
+
+    private:
+        // Sub-graph to be compiled and executed with MLIR.
+        const ngraph::op::CompiledKernel* m_compiledKernel;
+
+        // MLIR context that holds all the MLIR information related to the sub-graph
+        // compilation.
+        mlir::MLIRContext* m_context;
+        mlir::OpBuilder m_builder;
+
+        using TensorToInfo = std::pair<descriptor::Tensor*, TensorInfo>;
+        using TensorToInfoMap = std::unordered_map<descriptor::Tensor*, TensorInfo>;
+        using MLIRCompOpFunction = std::function<mlir::Operation*(
+            NgDialectConversionPass& NgDialectObj, const ngraph::Node*)>;
+        using MLIRCompOpMap = std::unordered_map<std::type_index, MLIRCompOpFunction>;
+
+        // Maps tensor to the value it represents in the IR
+        // use for MLIR dialect gen
+        TensorToInfoMap m_tensorToValueMap;
+        static const MLIRCompOpMap opDispatcher;
+    };
+
+} // end of namespace
+NgDialectConversionPass::NgDialectConversionPass(const NgDialectConversionPass& obj)
+    : m_compiledKernel(obj.m_compiledKernel)
+    , m_context(obj.m_context)
+    , m_builder(obj.m_builder)
+    , m_tensorToValueMap(obj.m_tensorToValueMap)
+{
+}
+
+void NgDialectConversionPass::runOnModule()
+{
+    TypeList argsTypeList, resultTypeList;
+
+    mlir::ModuleOp module = getModule();
+    // Retrieve input and output tensors.
+    const auto& kernelInputs = m_compiledKernel->get_arguments();
+    const auto& kernelOutput = m_compiledKernel->get_kernel_outputs();
+    NGRAPH_CHECK(kernelInputs.size() != 0, "Cannot have empty inputs list");
+    NGRAPH_CHECK(kernelOutput.size() != 0, "Cannot have empty outputs list");
+
+    for (auto input : kernelInputs)
+    {
+        argsTypeList.push_back(getMlirType(input.get()));
+    }
+
+    for (auto output : kernelOutput)
+    {
+        resultTypeList.push_back(getMlirType(output.get()));
+    }
+
+    auto funcType = mlir::FunctionType::get(argsTypeList, resultTypeList, m_context);
+    auto function = mlir::FuncOp::create(mlir::UnknownLoc::get(m_context), "main", funcType);
+    function.addEntryBlock();
+
+    // populate Tensor->Value maps
+    int i = 0;
+    for (auto input : kernelInputs)
+    {
+        mlir::Value* arg = function.getArgument(i);
+        TensorInfo tensorInfo{arg};
+        m_tensorToValueMap.insert(TensorToInfo(input->get_output_tensor_ptr().get(), tensorInfo));
+        i++;
+    }
+
+    // create builder
+    auto& region = function.getBody();
+    if (!region.empty())
+    {
+        m_builder.setInsertionPoint(&region.front(), region.front().begin());
```

Comment:
I think it is better to do buildNgDialect(function) and move the builder insertion point setting logic into buildNgDialect